src/core/connection_pool.c

```diff
@@ -0,0 +1,350 @@
+    CxPlatToeplitzHashInitialize(&ToeplitzHash);
+
+    Connections =
+        (HQUIC*)CXPLAT_ALLOC_PAGED(
+            sizeof(HQUIC) * NumberOfConnections,
+            QUIC_POOL_TMP_ALLOC);
+    if (Connections == NULL) {
+        Status = QUIC_STATUS_OUT_OF_MEMORY;
+        QuicTraceEvent(
+            AllocFailure,
+            "Allocation of '%s' failed. (%llu bytes)",
+            "Temp Connection Pool",
+            sizeof(HQUIC) * NumberOfConnections);
+        goto Error;
+    }
+
+    //
+    // Start testing ports and creating connections.
+    //
+
+    const uint32_t ConnectionsPerProc = (NumberOfConnections / RssProcessorCount) + 1;
+
+    for (uint32_t i = 0; i < NumberOfConnections; i++) {
+        uint32_t RssProcIndex;
+        do {
+            QuicAddrIncrementPort(&LocalAddress);
+            uint32_t RssHash = 0, Offset;
+            //
+            // Calculate the Toeplitz Hash as if receiving packets from the
+            // ResolvedRemoteAddress to find the RSS processor.
+            //
+            CxPlatToeplitzHashComputeRss(
+                &ToeplitzHash,
+                &ResolvedRemoteAddress,
+                &LocalAddress,
+                &RssHash,
+                &Offset);
+            uint32_t Mask = (RssConfig->RssIndirectionTableLength / sizeof(CXPLAT_PROCESSOR_INFO)) - 1;
+            RssProcIndex = RssProcessorCount;
+            CXPLAT_PROCESSOR_INFO* RssProc = &RssConfig->RssIndirectionTable[RssHash & Mask];
+            for (uint32_t j = 0; j < RssProcessorCount; j++) {
+                #pragma prefast(suppress:6385, "SAL doesn't understand this is safe");
+                if (RssProcessors[j].Group == RssProc->Group && RssProcessors[j].Index == RssProc->Index) {
+                    RssProcIndex = j;
+                    break;
+                }
+            }
+            CXPLAT_DBG_ASSERT(RssProcIndex < RssProcessorCount);
+            if (ConnectionCounts[RssProcIndex] >= ConnectionsPerProc) {
+                //
+                // This processor already has enough connections on it, so try another port number.
+                //
+                continue;
+            }
+
+            CXPLAT_SOCKET* TestSocket = NULL;
+            CXPLAT_UDP_CONFIG TestConfig = {
+                .LocalAddress = &LocalAddress,
+                .RemoteAddress = &ResolvedRemoteAddress
+            };
+            Status = CxPlatSocketCreateUdp(MsQuicLib.Datapath, &TestConfig, &TestSocket);
+            if (QUIC_FAILED(Status)) {
+                //
+                // This port is already in use. Gotta try another port number.
+                //
+                continue;
+            }
+
+            //
+            // Making it this far means we can create the connection!
+            //
+            CxPlatSocketDelete(TestSocket);
+            break;
+
+        } while (TRUE);
+
+        Status = MsQuicConnectionOpen(Registration, Handler, Context, &Connections[i]);
```

Comment:
Call `QuicConnAlloc` directly. Another thing: By default, creating a connection uses the "current partition" (i.e., current processor) to affinitize the connection, which isn't really what you want. You want it to be on the RSS partition. So, I would actually recommend modifying `QuicConnAlloc` to expose a new input parameter for `PartitionIndex` so you can set it explicitly.